### ISSUE TITLE
Dinit should diagnose deadlock when a service and its dependency both require console

### ISSUE THREAD

**Issue by mskarbek** [davmac314/dinit#6]
Simple situation - docker container with 3 services:
crond (`/etc/dinit.d/crond `):
```
type = process
command = /usr/sbin/crond -n -m off
restart = false
```
confd (`/etc/dinit.d/confd`):
```
type = scripted
command = /etc/dinit.d/confd.sh start
restart = false
options = runs-on-console
```
confd.sh (`/etc/dinit.d/confd.sh`):
```
#!/bin/sh

if [ "$1" = start ]; then
    /usr/bin/confd -onetime -backend etcd -node ${ETCD_HOST}
fi
```

nodejsapp (`/etc/dinit.d/nodejsapp`)
```
type = process
command = /usr/bin/node app.js
restart = false
options = runs-on-console
depends-ms = confd
start-timeout = 10
```

All that run by `boot`:
```
type = internal

waits-for = confd
waits-for = crond
waits-for = nodejsapp
```
Dinit fails to start `nodejsapp`. Container hangs at:
```
# docker exec -it nodejsapp bash
# dinitctl -s list
[{ }<<   ] boot
[{+}     ] confd
[{+}     ] crond
[{ }<<   ] nodejsapp
```
`confd` successfully finished it's job end exited with `0`, `crond` works fine but `nodejsapp` never starts:
```
# docker exec -it nodejsapp bash
# ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0  13724  3160 ?        Ss   16:32   0:00 /usr/sbin/dinit -s -d /etc/dinit.d
root         8  0.0  0.0  22964  2828 ?        Ss   16:32   0:00 /usr/sbin/crond -n -m off
root        19  0.0  0.0  12124  3188 pts/0    Ss   16:33   0:00 bash
root        49  0.0  0.0  34936  2992 pts/0    R+   16:45   0:00 ps aux
```
If you `exec` into container and manually stop and then start `nodejsapp` then it starts fine and have correct status in `dinitctl -s list`. `start-timeout = 10` seems to have no effect, container just hangs in that state for ever.

**Comment by davmac314** [comment]
I think the issue is that you have both `confd` and `nodejsapp` marked as `runs-on-console`. Since only one service can own the console at a time, the fact that `confd` holds it means it prevents `nodejsapp` from starting (it doesn't even begin the timeout, which is why setting that has no effect). Does the problem go away if you mark `confd` as `starts-on-console` instead of `runs-on-console`?

(As per the dinit-service manpage:  _A service running  on  the  console  prevents  other services from running on the console (they will queue for the console_))

That said, it sounds like a bug that stopping and starting `nodejsapp` does work, since it should just hit the same problem again. I'll have to investigate that.

**Comment by davmac314** [comment]
(Also, dinit should diagnose this kind of deadlock, but it doesn't do so yet).

**Comment by mskarbek** [comment]
As always RTFM to the end not only first few lines. :P After switching to `starts-on-console` container starts correctly. Thx!

**Comment by davmac314** [comment]
No problem. :)
Let's leave this issue open anyway, I will change the title: Dinit should diagnose deadlock when a service and its dependency both require console.

**Comment by davmac314** [comment]
"dinitctl list" now shows which service (if any) has and which are waiting for the console. I think that will probably do for this issue: it makes it much easier to diagnose what is happening, which was really the initial problem.

**Comment by davmac314** [comment]
Update dinitctl functionality included in 0.3.0 makes it easy to diagnose, so closing.
